File: hIndexII-LC 275.py
# Time Complexity = O(log n)
# Space Complexity = O(1)

class Solution:
    def hIndex(self, citations: list[int]) -> int:
        if len(citations) == 0:
            return 0
        
        n = len(citations)
        low = 0
        high = n - 1
        
        while(low <= high):
            mid = low + (high - low) // 2
            if citations[mid] >= n - mid:
                high = mid - 1          # Move left to find if there is a higher h index
            else:
                low = mid + 1           # Move right
                
        return n - low




# Time Complexity = O(n)
# Space Complexity = O(1)

class Solution:
    def hIndex(self, citations: list[int]) -> int:
        if len(citations) == 0:
            return 0
        
        n = len(citations)
        for i in range(n):
            if citations[i] >= n - i:
                return n - i
            
        
        return 0

File: H_IndexII_LC_275.java
/*
Did this code successfully run on Leetcode :  Yes
Any problem you faced while coding this :  No
Approach - Binary search
Time Complexity  - O(logN)
Space Complexity - O(1)
*/
public class H_IndexII_LC_275 {
        public int hIndex(int[] citations) {
            int start = 0;
            int end = citations.length-1;
            int n = citations.length;
            while(start <= end) {
                int mid =  start + (end-start) /2;
                if(citations[mid] == n-mid) return citations[mid];
                if(citations[mid] < n-mid) start = mid+1;
                if(citations[mid] > n-mid) end = mid-1;
            }
            return n-start;
        }
}


/* Linear
Time Complexity  - O(N)
Space Complexity - O(1)
 */
class Solution {
    public int hIndex(int[] citations) {
        int n = citations.length;
        for(int i = 0; i<n; i++) {
            int diff = n-i;
            if(diff <= citations[i]) {
                return diff;
            }
        }
        return 0;
    }
}


